Interface method decorator and core cleanup

Neverhorst opened the pull request:
## Description
This PR introduces a new concept to overload interface methods.
See new documentation _how_to_hardware_with_multiple_interfaces.md_ for more information on how to use this new feature.
There are some dummy modules for testing present which will be deleted before merging this PR into master:

- config/test.cfg
- hardware/myhardware.py
- interface/first_test_interface.py
- interface/second_test_interface.py

Also core folder/module structure has been organized a bit more in order to avoid circular dependencies and for consistency.

All this needs further testing since it impacts almost all modules.

**EDIT:** I also snuck in support for python native `frozenset` in the _yaml_ dumper/loader in order to support `frozenset` type StatusVars.

## Motivation and Context
When deriving a hardware module from multiple interfaces it is quite likely to run into problems with namespace collisions between these interface classes.
This PR enables you to define multiple implementations in a hardware module for a single interface method (occurring in multiple interfaces the modules inherits). The `Connector` object will hide this overload from logic modules, so nothing needs to be changed in those.

Many core classes have been moved to their own module and core paths have changed. Imports have been adjusted accordingly.

## How Has This Been Tested?
On Win8.1 x64 dummy and setup9
But needs more testing...

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:
- [x] My code follows the code style of this project.
- [x] I have documented my changes in the changelog (`documentation/changelog.md`)
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [ ] I have added/updated for the module the config example in the docstring of the class accordingly.
- [x] I have checked that the change does not contain obvious errors (syntax, indentation, mutable default values).
- [x] I have tested my changes using 'Load all modules' on the default dummy configuration with my changes included.
- [ ] All changed Jupyter notebooks have been stripped of their output cells.

kay-jahnke commented on core/interface.py:
```diff
@@ -64,11 +64,17 @@ def __get__(self, obj=None, cls=None):
+        if self.__isabstractmethod__:
+            raise NotImplementedError('No methods registered on abstractmethod {0}.'
+                                      ''.format(self.default_callable.__name__))
+        elif self.registered:
```
you are not checking the args or kwargs, how can you than determine that no keyword was given?

Neverhorst commented:
@kay-jahnke: yeah, the test logic got accidentally dumped too early since I want to get rid of all the testing stuff before merging.
If you want it you can check it out from an earlier version of this branch but it contained nothing of interest.

Neverhorst commented:
Ok, please check again. As we discussed I introduced more "draconic" constraints on the interface parameter in the `Connector` object.
Exceptions are raised when you try to connect to a module whose class name or base classes do not match `Connector.interface`.
I fixed all initialization calls for `Connector` in all modules within _default.cfg_ but this will potentially break any other module (especially modules connecting to non-interface classes)
